### PR TITLE
Add support to filter by last dagrun state in UI.

### LINKED ISSUE #42715: AIP-38 | DAGs List |  Filter by last_dag_run_state
We now have `last_dag_run_state` in FastAPI's get dags endpoint. Let's use that in the UI. And enable the disabled buttons that are already in the UI:

Image

### PR DESCRIPTION
**Pull request by tirkarthi**
closes: #42715
related: #42715

Add filter state on change to "lastrun" URL parameter which is used in legacy UI too and pass the state to dags list API.

### CODE REVIEW

**Comment by tirkarthi** [comment]
I am trying to run `pnpm lint:fix` which seems to give a different output compared to the errors in CI. Is there something missing in my local setup?

```
$ /home/karthikeyan/.local/share/pnpm/pnpm lint:fix

> ui@0.0.0 lint:fix /home/karthikeyan/stuff/python/airflow/airflow/ui
> eslint --fix && tsc --p tsconfig.app.json


/home/karthikeyan/stuff/python/airflow/airflow/ui/src/App.test.tsx
  35:20  warning  Use `undefined` instead of `null`  unicorn/no-null
  43:21  warning  Use `undefined` instead of `null`  unicorn/no-null
  45:21  warning  Use `undefined` instead of `null`  unicorn/no-null
  54:18  warning  Use `undefined` instead of `null`  unicorn/no-null
  55:23  warning  Use `undefined` instead of `null`  unicorn/no-null
  64:20  warning  Use `undefined` instead of `null`  unicorn/no-null
  72:21  warning  Use `undefined` instead of `null`  unicorn/no-null
  74:21  warning  Use `undefined` instead of `null`  unicorn/no-null
  83:18  warning  Use `undefined` instead of `null`  unicorn/no-null
  84:23  warning  Use `undefined` instead of `null`  unicorn/no-null

/home/karthikeyan/stuff/python/airflow/airflow/ui/src/pages/DagsList/DagsFilters.tsx
  30:28  warning  Arrow function has too many statements (12). Maximum allowed is 10  max-statements

/home/karthikeyan/stuff/python/airflow/airflow/ui/src/utils/index.tsx
  20:1   error  Expected a function expression                                                                             func-style
  27:10  error  Unexpected nullable string value in conditional. Please handle the nullish/empty cases explicitly          @typescript-eslint/strict-boolean-expressions
  27:39  error  Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`), as it is a safer operator  @typescript-eslint/prefer-nullish-coalescing

✖ 14 problems (3 errors, 11 warnings)

 ELIFECYCLE  Command failed with exit code 1.

```

**Comment by pierrejeambrun** [comment]
@tirkarthi

On my end, running `pnpm lint:fix` on your branch shows the same errors that the CI mentions.

It looks like your `eslint` is showing errors, where as the CI has no error on `eslint` part and shows errors from typescript only.

Can you verify your pnpm, eslint and node versions ?

**Comment by bbovenzi** [comment]
Yes, sometimes I need to make sure my local environment has the right version of node by running `nvm install 20 && nvm use 20`

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (pierrejeambrun) [inline airflow/ui/src/pages/DagsList/DagsFilters.tsx]:
```suggestion
const STATE_PARAM = "last_dag_run_state";
```

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (pierrejeambrun) [inline airflow/ui/src/pages/DagsList/DagsFilters.tsx]: I don't think we need to handle `all`. It is redundant with `no params set` I would say.

Anyway `all` can never end up in the `url` because it will just remove the query param from it. So i don't think we should check for `state === "all"`, only `state===null`

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (pierrejeambrun) [inline airflow/ui/src/pages/DagsList/DagsFilters.tsx]: null & undefined are !== anyway
```suggestion
  const isRunning = state === "running";
  const isFailed = state === "failed";
  const isSuccess = state === "success";
```

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (pierrejeambrun) [inline airflow/ui/src/pages/DagsList/DagsFilters.tsx:60]: Why a `React.MouseEventHandler` but `React.MouseEventHandler` above for the `pause`?

**Comment by bbovenzi** [comment]
> https://stackoverflow.com/questions/36836011/checking-validity-of-string-literal-union-type-at-runtime

We don't have an example of that yet. But I imagine we'll need a util like that. In the stackoverflow link. I would definitely lean towards a version of option 3.

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (pierrejeambrun) [inline airflow/ui/src/pages/DagsList/DagsFilters.tsx:68]: For the 'pause' we reset the display to 1st page. But here we don't do it.

Is there a reason for having an inconsistency of behavior between the filters ? I would expect to also reset the page to the first one.

**Review by pierrejeambrun — COMMENTED** [review]
Tested locally, working as expected.

A few minor suggestions :)

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (bbovenzi) [inline airflow/ui/src/pages/DagsList/DagsFilters.tsx]:
```suggestion
  const handleStateChange: React.MouseEventHandler<HTMLButtonElement> =
    useCallback(
      ({ currentTarget: { value } }) => {
```

This should remove the need to use `as`

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (tirkarthi) [reply airflow/ui/src/pages/DagsList/DagsFilters.tsx:60]: I initially had ChangeEventHandler but got below error and changed back to MouseEventHandler as suggested by the linter to keep it compatible with `onClick` .

```
const handleStateChange: React.ChangeEventHandler<HTMLButtonElement> =
```

Lint error : 

```
src/pages/DagsList/DagsFilters.tsx:85:15 - error TS2322: Type 'ChangeEventHandler<HTMLButtonElement>' is not assignable to type 'MouseEventHandler<HTMLButtonElement>'.
  Types of parameters 'event' and 'event' are incompatible.
    Type 'MouseEvent<HTMLButtonElement, MouseEvent>' is not assignable to type 'ChangeEvent<HTMLButtonElement>'.
      Types of property 'target' are incompatible.
        Type 'EventTarget' is not assignable to type 'EventTarget & HTMLButtonElement'.
          Type 'EventTarget' is missing the following properties from type 'HTMLButtonElement': disabled, form, formAction, formEnctype, and 311 more.

85               onClick={handleStateChange}
                 ~~~~~~~

  node_modules/.pnpm/@types+react@18.3.5/node_modules/@types/react/index.d.ts:2489:9
    2489         onClick?: MouseEventHandler<T> | undefined;
                 ~~~~~~~
    The expected type comes from property 'onClick' which is declared here on type 'IntrinsicAttributes & ButtonProps'
```

Comment (bbovenzi) [reply airflow/ui/src/pages/DagsList/DagsFilters.tsx:60]: Paused is a select which is a ChangeEvent while clicking on a button is just a MouseEvent

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (tirkarthi) [reply airflow/ui/src/pages/DagsList/DagsFilters.tsx]: I had kept it to `lastrun` for it to be compatible with legacy url once new UI becomes default in future and old links probably in bookmarks will continue to work. I will change it and don't have strong opinions too since users can update the bookmarks and links.

Comment (pierrejeambrun) [reply airflow/ui/src/pages/DagsList/DagsFilters.tsx]: I didn't have that in mind. Not sure if we want to try to be 'backward compatible'. I think we can take the airflow 3 opportunity to not have this extra constraint. Most of the fornt-end will change anyway, and I don't think we want to be contraints on `paths`, `params` etc.

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (pierrejeambrun) [reply airflow/ui/src/pages/DagsList/DagsFilters.tsx:60]: Got it thanks.

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (tirkarthi) [reply airflow/ui/src/pages/DagsList/DagsFilters.tsx:68]: Makes sense. Updated the code. Thanks.

File: airflow/ui/src/pages/DagsList/DagsFilters.tsx
Comment (bbovenzi) [reply airflow/ui/src/pages/DagsList/DagsFilters.tsx]: I want to leave us open to bolder redesigns so let's not worry about old->new url mapping for now. After the new UI is built we can add whatever redirects do still exist.

**Comment by bbovenzi** [comment]
> > > https://stackoverflow.com/questions/36836011/checking-validity-of-string-literal-union-type-at-runtime
> > 
> > 
> > We don't have an example of that yet. But I imagine we'll need a util like that. In the stackoverflow link. I would definitely lean towards a version of option 3.
> 
> @bbovenzi Quoting from the answer this needs the dag run states to be an array of strings. Since it's a type using union syntax I guess it's tricky. Please add in if I missed something or to change something since the part seems to be auto generated.
> 
> > In all cases you will need to store the strings in an object that is available at runtime (e.g. an array).

Perhaps something like this: https://stackoverflow.com/a/70694878

**Comment by tirkarthi** [comment]
> > > > https://stackoverflow.com/questions/36836011/checking-validity-of-string-literal-union-type-at-runtime
> > > 
> > > 
> > > We don't have an example of that yet. But I imagine we'll need a util like that. In the stackoverflow link. I would definitely lean towards a version of option 3.
> > 
> > 
> > @bbovenzi Quoting from the answer this needs the dag run states to be an array of strings. Since it's a type using union syntax I guess it's tricky. Please add in if I missed something or to change something since the part seems to be auto generated.
> > > In all cases you will need to store the strings in an object that is available at runtime (e.g. an array).
> 
> Perhaps something like this: https://stackoverflow.com/a/70694878

It seems it still needs specifying all union type values to be passed again to the util function as per the comment. I will leave it to another PR/issue to unblock this PR since the query parameter cannot be invalid from UI events unless user edits the URL intentionally or has a typo. It looks like a common request. Thanks for the pointers @bbovenzi .

**Comment by pierrejeambrun** [comment]
cc: @potiuk ☝️

**Review by bbovenzi — APPROVED** [review]

**Pull request merged**
